library/rsa.c

```diff
@@ -1523,6 +1523,112 @@ int mbedtls_rsa_rsassa_pss_sign( mbedtls_rsa_context *ctx,
+                                                 unsigned char *start,
+                                                 mbedtls_md_type_t md_alg,
+                                                 const unsigned char *hash,
+                                                 size_t hashlen )
+{
+    const mbedtls_md_info_t *md_info;
+    const char *oid;
+    size_t oid_size;
+
+    if( md_alg == MBEDTLS_MD_NONE )
+    {
+        if( *p < start + hashlen )
+            return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
+        *p -= hashlen;
+        memcpy( *p, hash, hashlen );
+        return( 0 );
+    }
+
+    md_info = mbedtls_md_info_from_type( md_alg );
+    if( md_info == NULL )
+        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
+    if( hashlen == 0 )
+        hashlen = mbedtls_md_get_size( md_info );
+    else if ( hashlen != mbedtls_md_get_size( md_info ) )
+        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
+    if( mbedtls_oid_get_oid_by_md( md_alg, &oid, &oid_size ) != 0 )
+        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
+
+    /* Double-check that 8 + hashlen + oid_size can be used as a
+     * 1-byte ASN.1 length encoding and that there's no overflow. */
+    if( 8 + hashlen + oid_size  >= 0x80         ||
+        10 + hashlen            <  hashlen      ||
+        10 + hashlen + oid_size <  10 + hashlen )
+        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
+
+    /*
+     * Static bounds check:
+     * - Need 10 bytes for five tag-length pairs.
+     *   (Insist on 1-byte length encodings to protect against variants of
+     *    Bleichenbacher's forgery attack against lax PKCS#1v1.5 verification)
+     * - Need hashlen bytes for hash
+     * - Need oid_size bytes for hash alg OID.
+     */
+    if( *p < start + 10 + oid_size + hashlen )
+        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
+    *p -= 10 + oid_size + hashlen;
+    start = *p;
+
+    /* Signing hashed data, add corresponding ASN.1 structure
+     *
+     * DigestInfo ::= SEQUENCE {
+     *   digestAlgorithm DigestAlgorithmIdentifier,
+     *   digest Digest }
+     * DigestAlgorithmIdentifier ::= AlgorithmIdentifier
+     * Digest ::= OCTET STRING
+     *
+     * Schematic:
+     * TAG-SEQ + LEN [ TAG-SEQ + LEN [ TAG-OID  + LEN [ OID  ]
+     *                                 TAG-NULL + LEN [ NULL ] ]
+     *                 TAG-OCTET + LEN [ HASH ] ]
+     */
+    *start++ = MBEDTLS_ASN1_SEQUENCE | MBEDTLS_ASN1_CONSTRUCTED;
+    *start++ = (unsigned char)( 0x08 + oid_size + hashlen );
+    *start++ = MBEDTLS_ASN1_SEQUENCE | MBEDTLS_ASN1_CONSTRUCTED;
+    *start++ = (unsigned char)( 0x04 + oid_size );
+    *start++ = MBEDTLS_ASN1_OID;
+    *start++ = (unsigned char) oid_size;
+    memcpy( start, oid, oid_size );
+    start += oid_size;
+    *start++ = MBEDTLS_ASN1_NULL;
+    *start++ = 0x00;
+    *start++ = MBEDTLS_ASN1_OCTET_STRING;
+    *start++ = (unsigned char) hashlen;
+    memcpy( start, hash, hashlen );
+    start += hashlen;
```

Comment:
This is redundant.

Reply:
Indeed but I prefer to retain the invariant that `start` points after what has been written and let the compiler optimize the addition away.